File: apps/studentWeb/src/contracts/pedagogicalContract.ts
export type PedagogicalArtifactType =
  | 'TEACH'
  | 'EXPLAIN'
  | 'SUMMARY'
  | 'MIND_MAP'
  | 'FLASHCARDS'
  | 'EXERCISES'
  | 'QUIZ'

export interface PedagogicalEvidenceContract {
  evidenceId: string | null
  materialId: string
  materialTitle: string
  locator: string
  excerpt: string
}

export interface PedagogicalArtifactContract {
  contractName: 'PedagogicalArtifact.v1'
  pedagogicalArtifactId: string
  studentId: string
  artifactType: PedagogicalArtifactType
  status: 'QUEUED' | 'RUNNING' | 'READY' | 'FAILED' | 'ARCHIVED'
  progressPercent: number
  message: string
  title: string
  instruction: string | null
  difficulty: string | null
  questionCount: number | null
  requestedTextModelId: string | null
  effectiveTextModelId: string | null
  thinkingMode: 'AUTO' | 'ON' | 'OFF'
  effectiveThinkingEnabled: boolean | null
  sourceMaterialIds: string[]
  sourceEvidence: PedagogicalEvidenceContract[]
  content: Record<string, unknown> | null
  errorCode: string | null
  errorMessage: string | null
  createdAt: string
  startedAt: string | null
  finishedAt: string | null
}

export interface LearningQuestionResultContract {
  questionId: string
  correct: boolean
  submittedAnswer: string
  correctAnswer: string
  explanation: string
}

export interface LearningAttemptContract {
  contractName: 'LearningAttempt.v1'
  learningAttemptId: string
  studentId: string
  pedagogicalArtifactId: string
  attemptType: 'EXERCISES' | 'QUIZ'
  scorePercent: number
  correctCount: number
  totalCount: number
  results: LearningQuestionResultContract[]
  adaptiveMessage: string
  updatedUnitIds: string[]
  createdAt: string
  completedAt: string
}

File: apps/backend/app/contracts/pedagogicalContract.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


PedagogicalArtifactType = Literal[
    "TEACH",
    "EXPLAIN",
    "SUMMARY",
    "MIND_MAP",
    "FLASHCARDS",
    "EXERCISES",
    "QUIZ",
]


class PedagogicalArtifactCreateContract(BaseModel):
    contractName: Literal[
        "PedagogicalArtifactCreate.v1"
    ] = "PedagogicalArtifactCreate.v1"

    artifactType: PedagogicalArtifactType
    studentLearningContextId: UUID | None = None
    studentSubjectId: UUID | None = None
    studentLearningUnitId: UUID | None = None
    title: str | None = Field(default=None, max_length=250)
    instruction: str | None = Field(default=None, max_length=2000)
    materialIds: list[UUID] = Field(default_factory=list)
    difficulty: Literal[
        "AUTO",
        "EASY",
        "MEDIUM",
        "HARD",
    ] = "AUTO"
    questionCount: int = Field(default=8, ge=1, le=30)
    requestedTextModelId: str | None = None
    thinkingMode: Literal["AUTO", "ON", "OFF"] = "AUTO"


class PedagogicalEvidenceContract(BaseModel):
    evidenceId: UUID | None
    materialId: UUID
    materialTitle: str
    locator: str
    excerpt: str


class PedagogicalArtifactContract(BaseModel):
    contractName: Literal[
        "PedagogicalArtifact.v1"
    ] = "PedagogicalArtifact.v1"

    pedagogicalArtifactId: UUID
    studentId: UUID
    studentLearningContextId: UUID | None
    studentSubjectId: UUID | None
    studentLearningUnitId: UUID | None
    artifactType: PedagogicalArtifactType
    status: Literal[
        "QUEUED",
        "RUNNING",
        "READY",
        "FAILED",
        "ARCHIVED",
    ]
    progressPercent: int
    message: str
    title: str
    instruction: str | None
    difficulty: str | None
    questionCount: int | None
    requestedTextModelId: str | None
    effectiveTextModelId: str | None
    thinkingMode: Literal["AUTO", "ON", "OFF"]
    effectiveThinkingEnabled: bool | None
    sourceMaterialIds: list[UUID]
    sourceEvidence: list[PedagogicalEvidenceContract]
    content: dict | None
    errorCode: str | None
    errorMessage: str | None
    createdAt: datetime
    startedAt: datetime | None
    finishedAt: datetime | None


class LearningAttemptSubmitContract(BaseModel):
    contractName: Literal[
        "LearningAttemptSubmit.v1"
    ] = "LearningAttemptSubmit.v1"

    answers: dict[str, str]


class LearningQuestionResultContract(BaseModel):
    questionId: str
    correct: bool
    submittedAnswer: str
    correctAnswer: str
    explanation: str


class LearningAttemptContract(BaseModel):
    contractName: Literal[
        "LearningAttempt.v1"
    ] = "LearningAttempt.v1"

    learningAttemptId: UUID
    studentId: UUID
    pedagogicalArtifactId: UUID
    attemptType: Literal["EXERCISES", "QUIZ"]
    scorePercent: int
    correctCount: int
    totalCount: int
    results: list[LearningQuestionResultContract]
    adaptiveMessage: str
    updatedUnitIds: list[UUID]
    createdAt: datetime
    completedAt: datetime
